mmseg/utils/util_distribution.py

```diff
@@ -0,0 +1,92 @@
+# Copyright (c) OpenMMLab. All rights reserved.
+import torch
+from mmcv.parallel import MMDataParallel, MMDistributedDataParallel
+
+dp_factory = {'cuda': MMDataParallel, 'cpu': MMDataParallel}
+ddp_factory = {'cuda': MMDistributedDataParallel}
+
+
+def build_dp(model, device='cuda', default_args=None):
+    """build DataParallel module by device type.
+
+    if device is cuda, return a MMDataParallel model; if device is mlu,
+    return a MLUDataParallel model.
+    Args:
+        model(nn.Module): model to be parallelized.
+        device(str): device type, ``'cuda'``, ``'cpu'`` or ``'mlu'``. Defaults to ``'cuda'``.
+        default_args(dict): including the following parameters.
+            device_ids(int): device ids of modules to be scattered to.
+                Defaults to None when GPU or MLU is not available.
+    Returns:
+        model(nn.Module): the model to be parallelized.
+    """
+    if device == 'cuda':
+        model = model.cuda()
+    elif device == 'mlu':
+        from mmcv.device.mlu import MLUDataParallel
+        dp_factory['mlu'] = MLUDataParallel
+        model = model.mlu()
+
+    return dp_factory[device](model, **default_args)
+
+
+def build_ddp(model, device='cuda', default_args=None):
+    """Build DistributedDataParallel module by device type.
+    
+    If device is cuda, return a MMDistributedDataParallel model; if device is
+    mlu, return a MLUDistributedDataParallel model.
+    
+    Args:
+        model(:class:`nn.Module`): module to be parallelized.
+        device(str): device type, mlu or cuda.
+        default_args(dict): including the following parameters.
+            device_ids(int): which represents the only device where the input
+                module corresponding to this process resides. Defaults to None.
+            broadcast_buffers(bool): Flag that enables syncing (broadcasting)
+                buffers of the module at beginning of the forward function.
+                Defaults to True.
+            find_unused_parameters(bool): Traverse the autograd graph of all
+                tensors contained in the return value of the wrapped module's
+                ``forward`` function. Parameters that don't receive gradients
+                as part of this graph are preemptively marked as being ready
+                to be reduced. Note that all ``forward`` outputs that are
+                derived from module parameters must participate in calculating
+                loss and later the gradient computation. If they don't, this
+                wrapper will hang waiting for autograd to produce gradients
+                for those parameters. Any outputs derived from module
+                parameters that are otherwise unused can be detached from the
+                autograd graph using ``torch.Tensor.detach``. Defaults to
+                False.
+    
+    Returns:
+        model(nn.Module): the module to be parallelized
```

Comment:
here should not be variable, just write type like
```suggestion
       nn.Module: the module to be parallelized
```